Yoast/wordpress-seo
# Add editor specific replace vars

## Summary

This PR can be summarized in the following changelog entry:

* Moves some snippet variables to only appear within specific editors. Adds a filter `wpseo_editor_specific_replace_vars` to make this pluggable.

## Relevant technical choices:

Due to time constraints I stopped with writing the tests and created a separate issue for finishing this: https://github.com/Yoast/wordpress-seo/issues/10180

* Editor specific replace vars is an array of page types, each with a list of replace var names.
* Add a filter `wpseo_editor_specific_replace_vars` to make this pluggable.
* Extract the `get_custom_fields` and `get_custom_taxonomies` to their own classes.
* Filters the replacement variables in `SettingsReplacementVariableEditors` so that each editor on the `settings -> search appearance` page receives only the relevant replacement variables.

## Test instructions

This PR can be tested by following these steps:

* Ensure all the snippet editors include the following replacement variables (these are shared):

Replacement variable | Nice name (English)
-- | --
date | Date
title | Title
archive_title | Archive title
sitename | Site title
excerpt | Excerpt
excerpt_only | Excerpt only
tag | Tag
category | Category
primary_category | Primary category
sep | Separator
modified | Modified
name | Name
user_description | User description
page | Page
pagetotal | Pagetotal
pagenumber | Pagenumber
caption | Caption
focuskw | Focus keyword

* Then for the following editors these variables are also there:

Page type (context) | Editor specific replacement variables
-- | --
post | id, pt_single, pt_plural, cf_<custom-field-name>, ct_<custom-tax-name>, ct_desc_<custom-tax-name>, term404
page | id, pt_single, pt_plural, cf_<custom-field-name>, parent_title
custom post type (custom-post-type) | id, pt_single, pt_plural, cf_<custom-field-name>, term404
category | category_description, term_description, term_title, parent_title
tag (post_tag) | tag_description, term_description, term_title
custom taxonomy (term-in-custom-taxonomy) | category_description, term_description, ct_<custom-tax-name>, ct_desc_<custom-tax-name>, parent_title
search | searchphrase


## Quality assurance


Fixes #10000
